### ISSUE TITLE
API: Custom objectives with zero hessian

### ISSUE THREAD

**Issue by DrumsnChocolate** [dmlc/xgboost#11057]
Posting here because the [discussion forum](https://discuss.xgboost.ai/) seems to be unavailable.

I'm using the Python API, and I've found that `reg:absoluteerror` works well for certain parts of the regression task I'm working on. However, there are also some unwanted characteristics that I think I can tackle with a custom objective. Similar to MAE, the hessian of my objective is zero. #7693 proposes to enable custom objectives with zero-hessians, but as far as I know that has not been realized yet.

I've tried to implement both squared error and absolute error in custom functions, to ensure that I know what I'm doing. While the custom implementation of the squared error perfectly mimics `reg:squarederror`, the custom absolute error does not match the behaviour of `reg:absoluteerror`. For reference, here are my custom functions that I pass to `xgboost.train`:
```python
def squared_error(y_pred, dtrain):
    y_true = dtrain.get_label()
    y_pred = y_pred.reshape(y_true.shape)
    grad = y_pred - y_true
    hess = np.ones_like(y_pred)
    return grad, hess


def absolute_error(y_pred, dtrain):
    y_true = dtrain.get_label()
    y_pred = y_pred.reshape(y_true.shape)
    grad = np.sign(y_pred - y_true)
    hess = np.ones_like(y_pred)
    return grad, hess
```

If you'd like, for completeness I can supply a script that shows that this `squared_error` exactly matches `reg:squarederror`, whereas `absolute_error` behaves differently from `reg:absoluteerror` (even though this function matches the way that gradients and hessians are computed in the c++ source I found here: https://github.com/dmlc/xgboost/blob/544a52e6ae438871024123b5de92b2314ddb3f78/src/objective/regression_obj.cu#L651-L661 )

Is there any progress with #7693, or is there a workaround I can use?

**Comment by trivialfis** [comment]
I haven't been able to work on that proposal yet. The `abs` error in XGBoost refreshes the tree leaf values based on MLE of samples in each leaf instead of relying solely on the gradient. For now, there's no way to do the same in Python.

**Comment by trivialfis** [comment]
I will close this in favor of https://github.com/dmlc/xgboost/issues/7693 . It's taking a bit of time, but the issue is still very much of interest. We do want to modularize it to make XGBoost more extensible, just a priority issue at the moment.
